### PR TITLE
ARROW-6613: [C++] Minimize usage of boost::filesystem

### PR DESCRIPTION
**Pull request by pitrou**
Allow building Arrow core (without tests, without filesystem layer) without boost dependencies.

The `minimal_build` example produces something like this:
```
# size  /usr/local/lib/libarrow.so
   text	   data	    bss	    dec	    hex	filename
4001919	  84992	   1712	4088623	 3e632f	/usr/local/lib/libarrow.so
# ldd  /usr/local/lib/libarrow.so
	linux-vdso.so.1 (0x00007fff45f40000)
	libstdc++.so.6 => /usr/lib/x86_64-linux-gnu/libstdc++.so.6 (0x00007f7089b08000)
	libm.so.6 => /lib/x86_64-linux-gnu/libm.so.6 (0x00007f708976a000)
	libgcc_s.so.1 => /lib/x86_64-linux-gnu/libgcc_s.so.1 (0x00007f7089552000)
	libpthread.so.0 => /lib/x86_64-linux-gnu/libpthread.so.0 (0x00007f7089333000)
	libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007f7088f42000)
	/lib64/ld-linux-x86-64.so.2 (0x00007f708a478000)
```

### CODE REVIEW

**Comment by pitrou** [comment]
Uh, looks like PyArrow is pulling in HDFS unconditionally.

File: cpp/CMakeLists.txt [hunk 1]
```diff
@@ -673,12 +673,19 @@ if(ARROW_STATIC_LINK_LIBS)
...
-set(ARROW_SHARED_PRIVATE_LINK_LIBS ${ARROW_STATIC_LINK_LIBS} ${BOOST_FILESYSTEM_LIBRARY}
-                                   ${BOOST_SYSTEM_LIBRARY})
+set(ARROW_SHARED_PRIVATE_LINK_LIBS ${ARROW_STATIC_LINK_LIBS})
 
-list(APPEND ARROW_STATIC_LINK_LIBS ${BOOST_FILESYSTEM_LIBRARY} ${BOOST_SYSTEM_LIBRARY})
+# Is boost::filesystem needed?
+if(ARROW_BUILD_INTEGRATION OR ARROW_BUILD_TESTS OR ARROW_FILESYSTEM OR ARROW_HDFS)
+  set(ARROW_WITH_BOOST_FILESYSTEM ON)
 
-list(APPEND ARROW_STATIC_INSTALL_INTERFACE_LIBS boost_filesystem boost_system boost_regex)
```
Comment (bkietz) [inline cpp/CMakeLists.txt]: could you comment on dropping boost_regex?

File: cpp/src/arrow/util/io_util.cc [hunk 2]
```diff
@@ -193,34 +210,79 @@ Status StringToNative(const std::string& s, NativePathString* out) {
...
+NativePathString NativeSlashes(NativePathString s) {
+  return ReplaceChars(std::move(s), kGenericSep, kNativeSep);
+}
+
+NativePathString GenericSlashes(NativePathString s) {
+  return ReplaceChars(std::move(s), kNativeSep, kGenericSep);
+}
+
+NativePathString NativeParent(const NativePathString& s) {
+  // XXX What about trailing separators?
```
Comment (bkietz) [inline cpp/src/arrow/util/io_util.cc]: Trailing separators should be ignored; `/a/b/` is equivalent to `/a/b` (and `/a//b//`). Also add tests for this

File: cpp/src/arrow/util/io_util.cc [hunk 3]
```diff
@@ -193,34 +210,79 @@ Status StringToNative(const std::string& s, NativePathString* out) {
...
-  std::uniform_int_distribution<int> dist(0, static_cast<int>(chars.length() - 1));
-
-  std::string s;
-  s.reserve(num_chars);
-  for (int i = 0; i < num_chars; ++i) {
-    s += chars[dist(gen)];
-  }
-  return s;
-}
+#endif
```
Comment (bkietz) [inline cpp/src/arrow/util/io_util.cc]: Would you mind decorating all multiline `#ifdef`/`#endif` blocks with comments?
```suggestion
#endif // ARROW_WITH_BOOST_FILESYSTEM
```

**Review by bkietz — CHANGES_REQUESTED** [review]

File: cpp/CMakeLists.txt (hunk 1 above)
Comment (pitrou) [reply cpp/CMakeLists.txt]: It's not used by Arrow, only by Parquet (and only internall I think?). @wesm

File: cpp/src/arrow/util/io_util.cc (hunk 3 above)
Comment (pitrou) [reply cpp/src/arrow/util/io_util.cc]: Will do.

File: cpp/src/arrow/util/io_util.cc (hunk 2 above)
Comment (pitrou) [reply cpp/src/arrow/util/io_util.cc]: Yes, will do.

**Comment by wesm** [comment]
I opened https://issues.apache.org/jira/browse/ARROW-6742, in case we care to remove the use of boost::filesystem in hdfs_internal.cc at some point. Use of HDFS is declining but probably not totally going away for another 10 years

**Comment by pitrou** [comment]
Also opened https://issues.apache.org/jira/browse/ARROW-6743 for other uses of `boost::filesystem`.

**Review by wesm — COMMENTED** [review]
Thanks for doing this. This is certainly quite tedious, but probably for the best in any case.

File: cpp/src/arrow/util/io_util.cc [hunk 4]
```diff
@@ -193,34 +210,89 @@ Status StringToNative(const std::string& s, NativePathString* out) {
...
 }
 
+#if _WIN32
+Status NativeToString(const NativePathString& ws, std::string* out) {
+#if _WIN32
+  std::string s;
+  RETURN_NOT_OK(::arrow::util::WideStringToUTF8(ws, &s));
+  *out = std::move(s);
+#else
+  *out = ws;
```
[the remark this answers is not in the document]
Comment (pitrou) [reply cpp/src/arrow/util/io_util.cc]: Uh, right. I had to `ifdef` out the whole function as it's only used on Windows.

**Comment by pitrou** [comment]
Note that `examples/minimal_build` still builds boost, even though it's not linked to. Header-only components of boost are used in various places.

**Comment by pitrou** [comment]
Rebased.

**Comment by pitrou** [comment]
Rebased.

**Comment by pitrou** [comment]
Rebased again in order to fix  AppVeyor failure.

File: cpp/src/arrow/util/io_util_test.cc [hunk 5]
```diff
@@ -77,6 +77,188 @@ TEST(PlatformFilename, Invalid) {
...
+  ASSERT_EQ(joined.ToNative(), L"a\\b\\c\\d");
+#else
+  ASSERT_EQ(joined.ToNative(), "a/b/c/d");
+#endif
+
+  ASSERT_OK(PlatformFilename::FromString("a/b/", &fn));
+  ASSERT_OK(fn.Join("c/d", &joined));
+  ASSERT_EQ(joined.ToString(), "a/b/c/d");
+#if _WIN32
+  ASSERT_EQ(joined.ToNative(), L"a\\b\\c\\d");
```
Comment (bkietz) [inline cpp/src/arrow/util/io_util_test.cc]: I'd really prefer if we could minimize the usage of `#if` like this. Could you instead define a helper which converts to the appropriate character type and joins segments with the platform's separator:
```suggestion
  ASSERT_EQ(joined.ToNative(), NativeJoin({"a", "b", "c", "d"}));
```

Comment (bkietz) [reply cpp/src/arrow/util/io_util_test.cc]: or even
```suggestion
  ASSERT_EQ(joined.ToNative(), FirstUnlessWindows("a/b/c/d", L"a\\b\\c\\d"));
```

Comment (bkietz) [reply cpp/src/arrow/util/io_util_test.cc]: I'll create a follow up JIRA for this.

**Review by bkietz — APPROVED** [review]

File: cpp/src/arrow/util/io_util_test.cc (hunk 5 above)
Comment (pitrou) [reply cpp/src/arrow/util/io_util_test.cc]: Ok, thanks. I like the explicit spelling, because it's clear what we are testing against.

**Comment by pitrou** [comment]
Green AppVeyor build: https://ci.appveyor.com/project/pitrou/arrow/builds/27859157

**Pull request closed without merging**